In [1]:
import numpy as np
import pandas as pd

# ==========================================
# INTEGRATION FUNCTIONS
# ==========================================

def trapezoidal_rule(func, a, b, n):
    """
    Approximates the integral of func from a to b using the Trapezoidal Rule.
    Formula: (h/2) * [f(a) + 2*sum(f(x_i)) + f(b)]
    """
    h = (b - a) / n
    
    # Sum of the first and last terms
    integration = func(a) + func(b)
    
    # Sum of the middle terms (multiplied by 2)
    for i in range(1, n):
        k = a + i * h
        integration += 2 * func(k)
        
    return integration * h / 2


def simpson_13_rule(func, a, b, n):
    """
    Approximates the integral using Simpson's 1/3 Rule.
    Requirement: n must be EVEN.
    Formula: (h/3) * [f(a) + 4*sum(odd) + 2*sum(even) + f(b)]
    """
    if n % 2 != 0:
        raise ValueError("For Simpson's 1/3 Rule, 'n' must be an even number.")
        
    h = (b - a) / n
    integration = func(a) + func(b)
    
    for i in range(1, n):
        k = a + i * h
        if i % 2 == 0:
            integration += 2 * func(k) # Even indices multiplied by 2
        else:
            integration += 4 * func(k) # Odd indices multiplied by 4
            
    return integration * h / 3


def simpson_38_rule(func, a, b, n):
    """
    Approximates the integral using Simpson's 3/8 Rule.
    Requirement: n must be a multiple of 3.
    Formula: (3h/8) * [f(a) + 3*sum(others) + 2*sum(multiples of 3) + f(b)]
    """
    if n % 3 != 0:
        raise ValueError("For Simpson's 3/8 Rule, 'n' must be a multiple of 3.")
        
    h = (b - a) / n
    integration = func(a) + func(b)
    
    for i in range(1, n):
        k = a + i * h
        if i % 3 == 0:
            integration += 2 * func(k) # Multiples of 3 multiplied by 2
        else:
            integration += 3 * func(k) # Others multiplied by 3
            
    return integration * 3 * h / 8

# ==========================================
# MAIN EXECUTION & TESTING
# ==========================================

# Define the function to integrate: f(x) = sin(x)
def f(x):
    return np.sin(x)

# Integration limits
a = 0
b = np.pi

# Number of intervals (n)
# Must be a multiple of both 2 (for 1/3 rule) and 3 (for 3/8 rule) to compare fairly.
# We choose n = 12.
n = 12

# 1. Exact Calculation
# Integral of sin(x) is -cos(x). evaluated from 0 to pi is -(-1) - (-1) = 2
exact_val = 2.0

# 2. Numerical Calculations
result_trap = trapezoidal_rule(f, a, b, n)
result_simp13 = simpson_13_rule(f, a, b, n)
result_simp38 = simpson_38_rule(f, a, b, n)

# 3. Compile Results
results_data = [
    ["Trapezoidal Rule", result_trap, abs(exact_val - result_trap)],
    ["Simpson's 1/3 Rule", result_simp13, abs(exact_val - result_simp13)],
    ["Simpson's 3/8 Rule", result_simp38, abs(exact_val - result_simp38)]
]

df_results = pd.DataFrame(results_data, columns=["Method", "Calculated Value", "Absolute Error"])

print(f"--- INTEGRATION OF sin(x) FROM 0 TO PI (n={n}) ---")
print(f"Exact Value: {exact_val}")
print("\n")
print(df_results.to_string(index=False))

--- INTEGRATION OF sin(x) FROM 0 TO PI (n=12) ---
Exact Value: 2.0


            Method  Calculated Value  Absolute Error
  Trapezoidal Rule          1.988564        0.011436
Simpson's 1/3 Rule          2.000053        0.000053
Simpson's 3/8 Rule          2.000119        0.000119


# Results and Discussion

### 1. Numerical Integration Results
We evaluated the integral $I = \int_{0}^{\pi} \sin(x) dx$ using three numerical methods with $n=12$ intervals. The exact analytical value of this integral is **2.0**.

| Method | Calculated Value | Absolute Error |
| :--- | :--- | :--- |
| **Trapezoidal Rule** | 1.9886 | 0.0114 |
| **Simpson's 1/3 Rule** | 2.0001 | 0.0001 |
| **Simpson's 3/8 Rule** | 2.0001 | 0.0001 |


### 2. Discussion
* **Trapezoidal Rule:** This method uses linear interpolation (straight lines) between points. While simple, it has the largest error among the three methods because a sine wave is curved, and straight lines cannot perfectly capture the area under the curve.
* **Simpson's Rules:** Both Simpson's 1/3 and 3/8 rules approximate the function using higher-order polynomials (parabolas for 1/3, cubic curves for 3/8). As seen in the results, their errors are significantly smaller (closer to zero), making them much more efficient for smooth functions like $\sin(x)$.